src/NuGet.Protocol.Catalog/CatalogClient.cs

```diff
@@ -98,34 +74,37 @@ public Task<CatalogLeaf> GetPackageDetailsLeafAsync(string leafUrl, Cancellation
             }
         }
 
-        private T DeserializeBytes<T>(byte[] jsonBytes)
+        private static T DeserializeBytes<T>(byte[] jsonBytes)
+            where T : class
         {
-            using (var stream = new MemoryStream(jsonBytes))
-            using (var textReader = new StreamReader(stream))
-            using (var jsonReader = new JsonTextReader(textReader))
+            using var stream = new MemoryStream(jsonBytes);
+            using var textReader = new StreamReader(stream);
+            using var jsonReader = new JsonTextReader(textReader);
+            var result = JsonSerializer.Deserialize<T>(jsonReader);
+            if (result == null)
             {
-                return JsonSerializer.Deserialize<T>(jsonReader);
+                throw new InvalidOperationException("Deserialization resulted in null");
             }
+
+            return result;
         }
 
         private async Task<T> DeserializeUrlAsync<T>(string documentUrl, CancellationToken token)
         {
             _logger.LogDebug("Downloading {documentUrl} as a stream.", documentUrl);
 
-            using (var response = await _httpClient.GetAsync(documentUrl, token))
-            using (var stream = await response.Content.ReadAsStreamAsync())
-            using (var textReader = new StreamReader(stream))
-            using (var jsonReader = new JsonTextReader(textReader))
+            using var response = await _httpClient.GetAsync(documentUrl, token);
+            await using var stream = await response.Content.ReadAsStreamAsync();
+            using var textReader = new StreamReader(stream);
+            using var jsonReader = new JsonTextReader(textReader);
+            try
             {
-                try
-                {
-                    return JsonSerializer.Deserialize<T>(jsonReader);
-                }
-                catch (JsonReaderException e)
-                {
-                    _logger.LogError(new EventId(0, documentUrl), e, "Failed to deserialize.");
-                    return default;
-                }
+                return JsonSerializer.Deserialize<T>(jsonReader);
```

Comment:
Last I tried (with Sentry's SDK) I hit some limitations. Since it wont' give us any clear benefit (a bit faster? this is a background worker and it's fast enough for what it does).
If we were doing it from scratch I'd give it a try

Reply:
Fair enough.